src/rewrite.jl

```diff
@@ -291,114 +291,279 @@ function _write_add_mul(vectorized, minus, current_sum, left_factors, inner_fact
+            # `x` so we cannot do `broadcast!(add_mul, I + x, 1)`.
+            code = Expr(:block)
+            if length(inner_factor.args) == 2
+                # Unary addition or subtraction.
                 next_sum = current_sum
                 start = 2
+            else
+                next_sum, new_code = _rewrite(
+                    vectorized,
+                    minus,
+                    inner_factor.args[2],
+                    current_sum,
+                    left_factors,
+                    right_factors,
+                )
+                push!(code.args, new_code)
+                start = 3
             end
-            vectorized = vectorized || inner_factor.args[1] == :(.+) || inner_factor.args[1] == :(.-)
             if inner_factor.args[1] == :- || inner_factor.args[1] == :(.-)
                 minus = !minus
             end
-            return rewrite_sum(vectorized, minus, inner_factor.args[start:end], next_sum, left_factors, right_factors, new_var, block)
+            vectorized = (
+                vectorized ||
+                inner_factor.args[1] == :(.+) ||
+                inner_factor.args[1] == :(.-)
+            )
+            return rewrite_sum(
+                vectorized,
+                minus,
+                inner_factor.args[start:end],
+                next_sum,
+                left_factors,
+                right_factors,
+                new_var,
+                code,
+            )
         elseif inner_factor.args[1] == :* && !vectorized
-            # We need `&& !vectorized` otherwise `x .+ A * b` would be rewritten `broadcast!(add_mul, x, A, b)`.
-
-            # we might need to recurse on multiple arguments, e.g.,
-            # (x+y)*(x+y)
-            # special case, only recurse on one argument and don't create temporary objects
-            if isone(mapreduce(_is_complex_expr, +, inner_factor.args)) &&
-               isone(mapreduce(_is_decomposable_with_factors, +, inner_factor.args))
-                # `findfirst` return the index in `2:...` so we need to add `1`.
-                which_idx = 1 + findfirst(2:length(inner_factor.args)) do i
-                    _is_decomposable_with_factors(inner_factor.args[i])
-                end
-                return _rewrite(
-                    vectorized, minus, inner_factor.args[which_idx], current_sum,
-                    vcat(left_factors, [esc(inner_factor.args[i]) for i in 2:(which_idx - 1)]),
-                    vcat(right_factors, [esc(inner_factor.args[i]) for i in length(inner_factor.args):-1:(which_idx + 1)]),
-                    new_var)
-            else
-                blk = Expr(:block)
+            # A multiplication expression *(args...). We need `!vectorized`
+            # otherwise `x .+ A * b` would be rewritten
+            # `broadcast!(add_mul, x, A, b)`.
+
+            # TODO(odow): remove the special case. I don't know if it's worth
```

Comment:
Why woudn't it be worth it, I don't feel comfortable doing this change, I would require careful benchmarking and testing with JuMP or a convincing argument :)

Reply:
Reinstated.